### ISSUE TITLE
String pointer miscalculation off by one

### ISSUE THREAD

**Issue by pcpa** [defeo/cypari2#46]
I am creating a python-cypari2 package for Fedora.
  I added this patch to prevent a crash on 32 bit arches, but the issue happens on all arches.
```
diff -up cypari2/gen.pyx.orig cypari2/gen.pyx
--- cypari2/gen.pyx.orig	2017-11-10 10:33:06.284726504 -0500
+++ cypari2/gen.pyx	2017-11-10 10:33:28.695726800 -0500
@@ -1617,7 +1617,7 @@ cdef class Gen(Gen_auto):
         lx = lgefint(x) - 2  # number of words
         size = lx * 4 * sizeof(long)
         s = <char *>check_malloc(size+3) # 1 char for sign, 1 char for 0, 1 char for '\0'
-        sp = s + size + 3
+        sp = s + size + 3 - 1
         sp[0] = 0
         xp = int_LSW(x)
         for i from 0 <= i < lx:
@@ -1660,7 +1660,7 @@ cdef class Gen(Gen_auto):
         lx = lgefint(x) - 2  # number of words
         size = lx*2*sizeof(long)
         s = <char *>check_malloc(size+4) # 1 char for sign, 2 chars for 0x, 1 char for '\0'
-        sp = s + size + 4
+        sp = s + size + 4 - 1
         sp[0] = 0
         xp = int_LSW(x)
         for i from 0 <= i < lx:
```

  There is also only one failure if running `tests/rundoctest.py` with python2 or python3.
  I believe it is just a side effect of 32 bit, so, for the moment, the check result is ignored.
  Example output:
```
+ PYTHONPATH=/builddir/build/BUILDROOT/python-cypari2-1.1.3-4.fc28.i386/usr/lib/python2.7/site-packages
+ export PATH PYTHONPATH
+ /usr/bin/python2 tests/rundoctest.py
  *** _^_: Warning: increasing stack size to 8192.
  *** _^_: Warning: increasing stack size to 16384.
  *** _^_: Warning: increasing stack size to 32768.
  *** _^_: Warning: increasing stack size to 65536.
  *** _^_: Warning: increasing stack size to 131072.
  *** _^_: Warning: increasing stack size to 262144.
  *** _^_: Warning: increasing stack size to 524288.
  *** _^_: Warning: increasing stack size to 1048576.
  *** _^_: Warning: increasing stack size to 2097152.
  *** _^_: Warning: increasing stack size to 4194304.
  *** _^_: Warning: increasing stack size to 8388608.
  *** _^_: Warning: increasing stack size to 16777216.
  ***   user warning: test
  *** _^s: Warning: increasing stack size to 2000000.
  *** _^s: Warning: increasing stack size to 4000000.
  *** _^s: Warning: increasing stack size to 8000000.
  *** _^s: Warning: increasing stack size to 16000000.
  ***   Warning: increasing stack size to 2048.
  ***   Warning: increasing stack size to 4096.
  ***   Warning: increasing stack size to 8192.
  ***   Warning: increasing stack size to 16384.
  ***   Warning: increasing stack size to 32768.
  ***   Warning: increasing stack size to 65536.
  ***   Warning: increasing stack size to 131072.
  ***   Warning: increasing stack size to 262144.
  ***   Warning: increasing stack size to 524288.
  ***   Warning: increasing stack size to 1048576.
  ***   Warning: increasing stack size to 2097152.
  ***   Warning: increasing stack size to 4194304.
  ***   Warning: increasing stack size to 8388608.
  ***   Warning: increasing stack size to 16777216.
================================================================================
Testing cypari2.closure
================================================================================
Testing cypari2.convert
================================================================================
Testing cypari2.gen
**********************************************************************
File "/builddir/build/BUILDROOT/python-cypari2-1.1.3-4.fc28.i386/usr/lib/python2.7/site-packages/cypari2/gen.so", line ?, in cypari2.gen.__test__.Gen.debug (line 4789)
Failed example:
    pari('[1/2, 1 + 1.0*I]').debug()
Differences (ndiff with -expected +actual):
    - [&=...] VEC(lg=3):...
    -   1st component = [&=...] FRAC(lg=3):...
    -     num = [&=...] INT(lg=3):... (+,lgefint=3):...
    -     den = [&=...] INT(lg=3):... (+,lgefint=3):...
    -   2nd component = [&=...] COMPLEX(lg=3):...
    -     real = [&=...] INT(lg=3):... (+,lgefint=3):...
    -     imag = [&=...] REAL(lg=3):... (+,expo=0):...
    + [&=5852aedc] VEC(lg=3):22000003 5852aed0 5852aeac
    +   1st component = [&=5852aed0] FRAC(lg=3):08000003 5852aec4 5852aeb8
    +     num = [&=5852aec4] INT(lg=3):02000003 (+,lgefint=3):40000003 00000001
    +     den = [&=5852aeb8] INT(lg=3):02000003 (+,lgefint=3):40000003 00000002
    +   2nd component = [&=5852aeac] COMPLEX(lg=3):0c000003 5852aea0 5852ae90
    +     real = [&=5852aea0] INT(lg=3):02000003 (+,lgefint=3):40000003 00000001
    +     imag = [&=5852ae90] REAL(lg=4):04000004 (+,expo=0):60000000 80000000 00000000
**********************************************************************
1 items had failures:
   1 of   3 in cypari2.gen.__test__.Gen.debug (line 4789)
***Test Failed*** 1 failures.
================================================================================
Testing cypari2.handle_error
================================================================================
Testing cypari2.pari_instance
================================================================================
Testing cypari2.stack
================================================================================
Testing cypari2.string_utils
================================================================================
Testing autogen.doc
================================================================================
Testing autogen.generator
================================================================================
Testing autogen.parser
================================================================================
Testing autogen.paths
================================================================================
Summary result for cypari2:
   attempted = 1367
   failed = 1
```

**Comment by videlec** [comment]
Thanks! The miscalculation is corrected in 540c4c8.

For the 32bits vs 64bits doctest error I do not see right now how it could be properly dealt with.

**Comment by videlec** [comment]
I am closing this issue. The doctest issue is now #47.
